copter: GCS failsafe now functions for all modes.

mday299 opened the pull request:
This patch causes the GCS failsafe to function when the aircraft is following waypoints in AUTO mode.  Previously the GCS failsafe only functioned for GUIDED mode.

khancyr commented on ArduCopter/events.cpp:
```diff
@@ -78,7 +78,7 @@ void Copter::failsafe_gcs_check()
 
     // return immediately if gcs failsafe is disabled, gcs has never been connected or we are not overriding rc controls from the gcs and we are not in guided mode
```
comment must be change too

khancyr commented:
Hi, have you test it ? I looks like to me that you only change the logic about rc_override 
With your patch we go from : 
" we are not overriding rc controls from the gcs and we are not in guided mode"
to
" we are not overriding rc controls from the gcs and we are not in guided mode and we are not in auto mode "
So I don't think this is right. Am I wrong ?

mday299 commented:
@khancyr -- I have tested this patch with SITL and it functions as I expect.

I have updated the comment as you indicated and updated the pull request.

magicrub commented:
Hi @mday299 the sitltest stage failed in Travis, can you take a look and see if it's from this code change? During the automated sim flight it disarmed and then stopped responding. I don't know if that's normal at that point or not, would need to compare to a working run. It may just be a fluke so I retriggered that test. We'll see..

OXINARF commented:
I think we need to be careful with this. It changes the default behavior of AUTO mode: if you connected a GCS and you lose that connection then it will enter RTL.
But I agree that the GCS failsafe needs to be reviewed for all modes so it doesn't depend on RC overrides being active.

magicrub commented:
Still failing Travis sim check. I agree that a gcs disconnect on the ground works rtl ir, and that's bad. Needs an is_flying check. For that matter, all failsafes need an is_flying check.

OXINARF commented:
@magicrub I'm not sure what you mean. If any failsafe engages in Copter and the vehicle is landed it will just disable the motors.

magicrub commented:
Gcs times out while on the ground and it didables motors?

OXINARF commented:
GCS failsafe can happen if you lose connection between the vehicle and GCS and:

- you are using RC overrides, OR
- you are in Guided mode
(with this PR, it adds another OR you are in Auto mode)

When the failsafe happens, if you are on the ground it disables motors, yes.

mday299 commented:
@magicrub I've finally run down the reason the why this causes Travis to fail.  Travis runs at 10X normal speed and the heartbeat from the GCS is checked every 5 seconds.  The GCS failsafe pretty much always fires when running the SITL this fast.  This didn't get seen before because the arducopter tests don't use GUIDED mode (and the GCS failsafe is only active for GUIDED mode).  

Running this patch through the tests at normal speed does not trigger the same problem.

magicrub commented:
Hmm... That's odd. I guess you've uncovered a bug in our Travis SITL test!

mday299 commented:
@rmackay9 -- I'd be curious to know your thoughts on this PR.  I'd also be curious as to what to do about the failing Travis test.

OXINARF commented:
@mday299 This was talked yesterday in the dev call. Randy is fine with it, although we talked about a more generic approach than just enabling it for Auto.
The issue with autotest will need to be fixed in MAVProxy and until then this can't get go in otherwise breaks our CI testing.

Also I'd prefer this *if* is split, it is getting very hard to read. And I think the default for FS_GCS_ENABLE should be changed to continue mission (keeping the same behavior as now), but we'll confirm this with Randy.

mday299 commented:
@OXINARF If the default FS_GCS_ENABLE is changed to continue mission that would very likely keep the CI working.  I'd be curious to know if @rmackay9 is OK with using that new default.  If so, I could work on splitting up the "if" statement and also applying the GCS check to modes other than AUTO and GUIDED. Which modes are wanted?

OXINARF commented:
@mday299 We didn't discuss in detail but I think the general idea is to have it working in every mode. A real GCS failsafe. So it would trigger if:

- using RC overrides and it times out OR
- enabled and connection from GCS times out

This would probably require new options for, at least, _enabled only in Guided_ and _enabled only for Guided and Auto_.

mday299 commented:
@OXINARF I updated the pull request.  GCS failsafe triggers based on the criteria you outlined.  Added two more options to the FS_GCS_ENABLE parameter: FS_GCS_ENABLED_GUIDED, FS_GCS_ENABLED_GUIDED_OR_AUTO.  Made the default for FS_GCS_ENABLE = FS_GCS_ENABLED_CONTINUE_MISSION.  

Also added a paramter: FS_GCS_TIMEOUT that governs how long to wait after loss of heartbeat from GCS before triggering the failsafe.

To get the autotest stuff working here locally I had to also add "FS_GCS_ENABLE 2" to the Tools/autotest/default_params/copter.param file.  However, Travis doesn't use the corresponding file from my branch when testing, so the FS_GCS_ENALBE param is not correct and CI still fails.  I think if copter.parm is updated for master then Travis will not fail with this PR.

What do you think?

magicrub commented on ArduCopter/Parameters.h:
```diff
@@ -292,7 +292,9 @@ class Parameters {
         k_param_radio_tuning,
         k_param_radio_tuning_high,
         k_param_radio_tuning_low,
-        k_param_rc_speed = 192,
```
I don't like this switch-a-roo even though it's technically correct. Can't you just insert the new k_param_failsafe_gcs_timeout below k_param_radio_tuning_low? Currently k_param_radio_tuning_low == 189 and k_param_rc_speed == 192, this new one should be k_param_failsafe_gcs_timeout == 190 and since there's one empty spot there hiding then it would be nice to explicitly state "k_param_failsafe_gcs_timeout = 190," to reduce future human error.

magicrub commented on ArduCopter/events.cpp:
```diff
@@ -87,7 +87,7 @@ void Copter::failsafe_gcs_check()
 
     // check if all is well
-    if (last_gcs_update_ms < FS_GCS_TIMEOUT_MS) {
+    if (last_gcs_update_ms < (uint8_t) g.failsafe_gcs_timeout * 1000) {
```
this casting is unintuitive, can we cast to uint32_t instead

OXINARF commented on ArduCopter/events.cpp:
```diff
@@ -113,10 +113,25 @@ void Copter::failsafe_gcs_check()
+            (control_mode == GUIDED && g.failsafe_gcs == FS_GCS_ENABLED_GUIDED) ||
+            ((control_mode == GUIDED || control_mode == AUTO) && g.failsafe_gcs == FS_GCS_ENABLED_GUIDED_OR_AUTO)) {
+                gcs_send_text(MAV_SEVERITY_ERROR,"Lost contact with GCS!");
+                set_mode_RTL_or_land_with_pause(MODE_REASON_GCS_FAILSAFE);
```
I'd like this entire else to be something like:

```
switch (g.failsafe_gcs) {
case FS_GCS_ENABLED_CONTINUE_MISSION:
  if (control_mode == AUTO) {
    set_mode_RTL...;
  }
  break;

case FS_GCS_ENABLED_GUIDED:
  if (control_mode != GUIDED) {
    break;
  }

case FS_GCS_ENABLED_GUIDED_OR_AUTO:
  if (control_mode != GUIDED && control_mode != AUTO) {
    break;
  }

case FS_GCS_ENABLED_ALWAYS_RTL: 
  set_mode_RTL...;
  break;
}
```

I think this is much more readable with only repeating the call to set mode once. Regarding the text to the GCS, I agree with it, but it should be above after the *Log_Write_Error* call and I think that it should be a warning, not an error (unless you have found another failsafe where an error is used).

OXINARF commented on ArduCopter/defines.h:
```diff
@@ -469,6 +469,8 @@ enum DevOptions {
 #define FS_GCS_ENABLED_ALWAYS_RTL           1
 #define FS_GCS_ENABLED_CONTINUE_MISSION     2
+#define FS_GCS_ENABLED_GUIDED               3
+#define FS_GCS_ENABLED_GUIDED_OR_AUTO       4
```
Not really important but I would remove the OR.

mday299 commented:
@magicrub and @OXINARF -- I implemented the requested changes but I had to tweak in a couple spots to get things working.

CI checks will fail due to the RC overrides triggering (coupled with the previously identified problem with running at 10x real time speed).  Dunno what to do about that.

mday299 commented on ArduCopter/events.cpp:
```diff
@@ -105,19 +112,41 @@ void Copter::failsafe_gcs_check()
+            default:
+                gcs_send_text(MAV_SEVERITY_WARNING,
+                        "Unrecognized GCS failsafe enum");
+                break;
```
Good point. I'll change it to always RTL here.

OXINARF commented:
@mday299 Hopefully I'm on time before you make the changes. We had an interesting discussion last night in Gitter. Randy brought the good point that RC override is already covered by the throttle failsafe so we decided to completely remove the special cases for RC override here, it should just check the last heartbeat variable without caring if RC override is active or not.

We also discussed what the default should be. Although I was more inclined to have it enabled only for Guided and Auto, I'm more convinced now to enable it for all modes. The failsafe won't happen if a GCS is never connected and you inserted a parameter to control the timeout so that should cover most cases where it could annoy people. Sorry to change this again, but let's go back to enabled for all modes.

Sorry for being a pain. Clearly this wasn't done before as people have strong opinions about it.

mday299 commented:
@OXINARF -- made the changes you requested in the PRs.  Then removed RC override and changed GCS default.  

Obviously with the new default the CI REALLY won't work. :)

mday299 commented:
@OXINARF and @magicrub --bump.  :)

OXINARF commented:
@mday299 This is OK. I've cleaned up the parameter descriptions a bit and some white-space you left and will push that soon in my fork for you to look at. I've also done the support for SITL speedup in MAVProxy but it still needs more testing (and needs to go in and be released).

magicrub commented on ArduCopter/events.cpp:
```diff
@@ -105,18 +105,38 @@ void Copter::failsafe_gcs_check()
+            case FS_GCS_ENABLED_GUIDED:
+                if (control_mode != GUIDED) {
+                    break;
+                }
```
these two fall throughs need to be explicitly commented to prevent misinterpretation and someone "fixing" it later

// FALL THROUGH

magicrub commented:
@OXINARF there was talk at one point to make this work on SITL, was there a change put in that would work if this was rebased on master? or does MAVproxy need to be changed to respond to heartbeats when received?

OXINARF commented:
@magicrub Like I said above, I've done the work for MAVProxy (and also changes in SITL) but it needs polishing as Quadplane failed the tests.

mday299 commented:
@OXINARF have you had a chance to polish up the MAVProxy quadplane issue? 

BTW, Thanks for all the help on this!

OXINARF commented:
@mday299 I started reworking it on Friday, but couldn't finish it. I'll try to do it today, but I'm not sure if I'm going to have time.

mday299 commented:
@OXINARF have you had a chance to look at this any more?

OXINARF commented:
@mday299 I did, but it wasn't working yet. I'm close but haven't had much time to work on it.

mday299 commented:
@OXINARF -- bump :)

OXINARF commented:
@mday299 I'm so sorry, it's not forgotten, but my time has been very limited for any code development.

Naterater commented:
So this works just fine when following waypoints in Plane. Can I ask what is the basis of the holdup? Trying to close some issues, especially https://github.com/ArduPilot/ardupilot/issues/705

amilcarlucas commented:
Needs a rebase, and probably will clash with the FS_OPTIONS PR that Matt has been working on.

Pedals2Paddles commented:
Closed by https://github.com/ArduPilot/ardupilot/pull/7432
